server/src/main/java/io/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -157,53 +161,83 @@ public void stop()
     }
   }
 
+  private <T> T inReadOnlyTransaction(final TransactionCallback<T> callback)
+  {
+    return connector.getDBI().withHandle(
+        new HandleCallback<T>()
+        {
+          @Override
+          public T withHandle(Handle handle) throws Exception
+          {
+            final Connection connection = handle.getConnection();
+            final boolean readOnly = connection.isReadOnly();
+            connection.setReadOnly(true);
+            try {
+              T res = handle.inTransaction(callback);
+              connection.setReadOnly(readOnly);
+              return res;
+            } catch(TransactionFailedException e) {
+              connection.setReadOnly(readOnly);
+              throw Throwables.propagate(e);
```

Comment:
possibly, then we swallow the exception on setReadOnly, but maybe it's better than not propagating the transaction failure.